File: app/src/main/java/pl/piotrskiba/angularowo/fragments/OffersFragment.java
package pl.piotrskiba.angularowo.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.reward.RewardedVideoAd;

import butterknife.BindView;
import butterknife.ButterKnife;
import pl.piotrskiba.angularowo.AppViewModel;
import pl.piotrskiba.angularowo.R;
import pl.piotrskiba.angularowo.adapters.AdOffersAdapter;
import pl.piotrskiba.angularowo.adapters.OffersAdapter;
import pl.piotrskiba.angularowo.interfaces.AdOfferClickListener;
import pl.piotrskiba.angularowo.interfaces.NetworkErrorListener;
import pl.piotrskiba.angularowo.interfaces.OfferClickListener;
import pl.piotrskiba.angularowo.models.AdOffer;
import pl.piotrskiba.angularowo.models.Offer;
import pl.piotrskiba.angularowo.models.OffersInfo;
import pl.piotrskiba.angularowo.network.ServerAPIClient;
import pl.piotrskiba.angularowo.network.ServerAPIInterface;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class OffersFragment extends Fragment implements AdOfferClickListener, OfferClickListener, NetworkErrorListener {

    @BindView(R.id.swiperefresh)
    SwipeRefreshLayout mSwipeRefreshLayout;

    @BindView(R.id.tv_coins)
    TextView mCoinsTextView;

    @BindView(R.id.rv_ad_offers)
    RecyclerView mAdOffersRecyclerView;

    @BindView(R.id.rv_offers)
    RecyclerView mOffersRecyclerView;

    @BindView(R.id.noRanksAvailableTextView)
    TextView mNoRanksAvailableTextView;

    @BindView(R.id.no_internet_layout)
    LinearLayout mNoInternetLayout;

    @BindView(R.id.server_error_layout)
    LinearLayout mServerErrorLayout;

    @BindView(R.id.account_banned_layout)
    LinearLayout mAccountBannedLayout;

    @BindView(R.id.default_layout)
    NestedScrollView mDefaultLayout;

    private RewardedVideoAd mRewardedVideoAd;

    private OffersInfo mOffersInfo;
    private AdOffersAdapter mAdOffersAdapter;
    private OffersAdapter mOffersAdapter;

    public OffersFragment(){

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_offers, container, false);

        ButterKnife.bind(this, view);

        setupAdOffersRecyclerView();
        setupOffersRecyclerView();

        seekForOffersInfo();

        ActionBar actionbar = ((AppCompatActivity)getActivity()).getSupportActionBar();
        actionbar.setTitle(R.string.actionbar_title_free_ranks);

        mSwipeRefreshLayout.setOnRefreshListener(() -> refreshData());

        return view;
    }

    private void setupAdOffersRecyclerView(){
        mAdOffersAdapter = new AdOffersAdapter(getContext(), this);

        RecyclerView.LayoutManager adOffersLayoutManager;
        adOffersLayoutManager = new GridLayoutManager(getContext(), 4);

        mAdOffersRecyclerView.setAdapter(mAdOffersAdapter);
        mAdOffersRecyclerView.setLayoutManager(adOffersLayoutManager);
        mAdOffersRecyclerView.setHasFixedSize(true);
        mAdOffersRecyclerView.setNestedScrollingEnabled(false);
    }

    private void setupOffersRecyclerView(){
        mOffersAdapter = new OffersAdapter(getContext(), this);

        RecyclerView.LayoutManager offersLayoutManager;
        offersLayoutManager = new GridLayoutManager(getContext(), 3);

        mOffersRecyclerView.setAdapter(mOffersAdapter);
        mOffersRecyclerView.setLayoutManager(offersLayoutManager);
        mOffersRecyclerView.setHasFixedSize(true);
        mOffersRecyclerView.setNestedScrollingEnabled(false);
    }

    @Override
    public void onAdOfferClick(View view, AdOffer clickedAdOffer) {
        if(getContext() != null && !mSwipeRefreshLayout.isRefreshing() && clickedAdOffer.getTimeleft() == 0) {
            new AlertDialog.Builder(getContext())
                    .setTitle(R.string.ad_question)
                    .setMessage(R.string.ad_question_description)
                    .setPositiveButton(R.string.button_yes, (dialogInterface, i) -> {
                        loadRewardedVideoAd(clickedAdOffer.getAdId());
                    })
                    .setNegativeButton(R.string.button_no, null)
                    .show();
        }
    }

    @Override
    public void onOfferClick(View view, Offer clickedOffer) {
        if(getContext() != null && !mSwipeRefreshLayout.isRefreshing() && clickedOffer.getTimeleft() == 0 && clickedOffer.getPrice() <= mOffersInfo.getPoints()) {
            new AlertDialog.Builder(getContext())
                    .setTitle(R.string.offer_question)
                    .setMessage(getResources().getQuantityString(R.plurals.offer_question_description, clickedOffer.getPrice(), clickedOffer.getTitle(), clickedOffer.getPrice()))
                    .setPositiveButton(R.string.button_yes, (dialogInterface, i) -> {
                        redeemOffer(clickedOffer);
                    })
                    .setNegativeButton(R.string.button_no, null)
                    .show();
        }
    }

    private void seekForOffersInfo(){
        AppViewModel viewModel = new ViewModelProvider(requireActivity()).get(AppViewModel.class);
        viewModel.setNetworkErrorListener(this);
        viewModel.getOffersInfo().observe(this, offersInfo -> {
            if(offersInfo != null){
                mOffersInfo = offersInfo;
                populateUi();
            }
        });
    }

    public void refreshData(){
        mSwipeRefreshLayout.setRefreshing(true);

        AppViewModel viewModel = new ViewModelProvider(requireActivity()).get(AppViewModel.class);
        viewModel.refreshOffersInfo();
    }

    private void populateUi(){
        showDefaultLayout();

        mCoinsTextView.setText(String.valueOf(mOffersInfo.getPoints()));

        mAdOffersAdapter.setAdOfferList(mOffersInfo.getAdOffers());
        mOffersAdapter.setOfferList(mOffersInfo.getOffers(), mOffersInfo.getPoints());
    }

    private void loadRewardedVideoAd(String adId) {
        showLoadingIndicator();
        mRewardedVideoAd.loadAd(adId, new AdRequest.Builder().build());
    }

    private void redeemOffer(Offer offer){
        showLoadingIndicator();

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getContext());
        String access_token = sharedPreferences.getString(getString(R.string.pref_key_access_token), null);

        Context context = getContext();

        ServerAPIInterface serverAPIInterface = ServerAPIClient.getRetrofitInstance().create(ServerAPIInterface.class);
        serverAPIInterface.redeemOffer(ServerAPIClient.API_KEY, offer.getId(), access_token).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if(isAdded()) {
                    new AlertDialog.Builder(context)
                            .setTitle(R.string.offer_redeemed)
                            .setMessage(R.string.offer_redeemed_description)

                            .setPositiveButton(R.string.button_dismiss, null)
                            .show();

                    refreshData();
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void showDefaultLayout(){
        hideLoadingIndicator();
        mNoRanksAvailableTextView.setVisibility(View.GONE);
        mDefaultLayout.setVisibility(View.VISIBLE);
        mNoInternetLayout.setVisibility(View.GONE);
        mServerErrorLayout.setVisibility(View.GONE);
        mAccountBannedLayout.setVisibility(View.GONE);
    }
    private void showNoRewardsLayout(){
        hideLoadingIndicator();
        mNoRanksAvailableTextView.setVisibility(View.VISIBLE);
        mDefaultLayout.setVisibility(View.GONE);
        mNoInternetLayout.setVisibility(View.GONE);
        mServerErrorLayout.setVisibility(View.GONE);
        mAccountBannedLayout.setVisibility(View.GONE);
    }
    private void showNoInternetLayout(){
        hideLoadingIndicator();
        mNoRanksAvailableTextView.setVisibility(View.GONE);
        mDefaultLayout.setVisibility(View.GONE);
        mNoInternetLayout.setVisibility(View.VISIBLE);
        mServerErrorLayout.setVisibility(View.GONE);
        mAccountBannedLayout.setVisibility(View.GONE);
    }
    private void showServerErrorLayout(){
        hideLoadingIndicator();
        mNoRanksAvailableTextView.setVisibility(View.GONE);
        mDefaultLayout.setVisibility(View.GONE);
        mNoInternetLayout.setVisibility(View.GONE);
        mServerErrorLayout.setVisibility(View.VISIBLE);
        mAccountBannedLayout.setVisibility(View.GONE);
    }
    private void showAccountBannedLayout(){
        hideLoadingIndicator();
        mNoRanksAvailableTextView.setVisibility(View.GONE);
        mDefaultLayout.setVisibility(View.GONE);
        mNoInternetLayout.setVisibility(View.GONE);
        mServerErrorLayout.setVisibility(View.GONE);
        mAccountBannedLayout.setVisibility(View.VISIBLE);
    }

    private void showLoadingIndicator(){
        mSwipeRefreshLayout.setRefreshing(true);
    }
    public void hideLoadingIndicator(){
        mSwipeRefreshLayout.setRefreshing(false);
    }

    public void setRewardedVideoAd(RewardedVideoAd ad){
        mRewardedVideoAd = ad;
    }

    @Override
    public void onNoInternet() {
        showNoInternetLayout();
    }

    @Override
    public void onServerError() {
        showServerErrorLayout();
    }
}

File: app/src/main/java/pl/piotrskiba/angularowo/fragments/OffersFragment.kt
package pl.piotrskiba.angularowo.fragments

import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.NestedScrollView
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import butterknife.BindView
import butterknife.ButterKnife
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.reward.RewardedVideoAd
import pl.piotrskiba.angularowo.AppViewModel
import pl.piotrskiba.angularowo.R
import pl.piotrskiba.angularowo.adapters.AdOffersAdapter
import pl.piotrskiba.angularowo.adapters.OffersAdapter
import pl.piotrskiba.angularowo.fragments.base.BaseFragment
import pl.piotrskiba.angularowo.interfaces.AdOfferClickListener
import pl.piotrskiba.angularowo.interfaces.NetworkErrorListener
import pl.piotrskiba.angularowo.interfaces.OfferClickListener
import pl.piotrskiba.angularowo.models.AdOffer
import pl.piotrskiba.angularowo.models.DetailedPlayer
import pl.piotrskiba.angularowo.models.Offer
import pl.piotrskiba.angularowo.models.OffersInfo
import pl.piotrskiba.angularowo.network.ServerAPIClient
import pl.piotrskiba.angularowo.network.ServerAPIClient.retrofitInstance
import pl.piotrskiba.angularowo.network.ServerAPIInterface
import pl.piotrskiba.angularowo.utils.AnalyticsUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class OffersFragment : BaseFragment(), AdOfferClickListener, OfferClickListener, NetworkErrorListener {

    private lateinit var mViewModel: AppViewModel
    private lateinit var mRewardedVideoAd: RewardedVideoAd
    private lateinit var mOffersInfo: OffersInfo
    private lateinit var mAdOffersAdapter: AdOffersAdapter
    private lateinit var mOffersAdapter: OffersAdapter
    private var mPlayer: DetailedPlayer? = null

    @BindView(R.id.swiperefresh)
    lateinit var mSwipeRefreshLayout: SwipeRefreshLayout

    @BindView(R.id.tv_coins)
    lateinit var mCoinsTextView: TextView

    @BindView(R.id.rv_ad_offers)
    lateinit var mAdOffersRecyclerView: RecyclerView

    @BindView(R.id.tv_limit_reached)
    lateinit var mOfferLimitReachedTextView: TextView

    @BindView(R.id.rv_offers)
    lateinit var mOffersRecyclerView: RecyclerView

    @BindView(R.id.noRanksAvailableTextView)
    lateinit var mNoRanksAvailableTextView: TextView

    @BindView(R.id.no_internet_layout)
    lateinit var mNoInternetLayout: LinearLayout

    @BindView(R.id.server_error_layout)
    lateinit var mServerErrorLayout: LinearLayout

    @BindView(R.id.account_banned_layout)
    lateinit var mAccountBannedLayout: LinearLayout

    @BindView(R.id.default_layout)
    lateinit var mDefaultLayout: NestedScrollView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mViewModel = ViewModelProvider(requireActivity()).get(AppViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_offers, container, false)

        ButterKnife.bind(this, view)

        setupAdOffersRecyclerView()
        setupOffersRecyclerView()
        seekForPlayerUpdates()
        seekForOffersInfo()

        val actionbar = (activity as AppCompatActivity?)?.supportActionBar
        actionbar?.setTitle(R.string.actionbar_title_free_ranks)

        mSwipeRefreshLayout.setOnRefreshListener { refreshData() }

        return view
    }

    private fun setupAdOffersRecyclerView() {
        mAdOffersAdapter = AdOffersAdapter(requireContext(), this)
        mAdOffersRecyclerView.adapter = mAdOffersAdapter

        val adOffersLayoutManager = GridLayoutManager(context, resources.getInteger(R.integer.ad_offer_list_span_count))
        mAdOffersRecyclerView.layoutManager = adOffersLayoutManager

        mAdOffersRecyclerView.setHasFixedSize(true)
        mAdOffersRecyclerView.isNestedScrollingEnabled = false
    }

    private fun setupOffersRecyclerView() {
        mOffersAdapter = OffersAdapter(requireContext(), this)
        mOffersRecyclerView.adapter = mOffersAdapter

        val offersLayoutManager = GridLayoutManager(context, resources.getInteger(R.integer.offer_list_span_count))
        mOffersRecyclerView.layoutManager = offersLayoutManager

        mOffersRecyclerView.setHasFixedSize(true)
        mOffersRecyclerView.isNestedScrollingEnabled = false
    }

    override fun onAdOfferClick(view: View, clickedAdOffer: AdOffer) {
        if (context != null && !mSwipeRefreshLayout.isRefreshing && clickedAdOffer.timeleft == 0) {
            AnalyticsUtils().logAdOfferDialogOpen(
                    mPlayer?.uuid ?: "",
                    mPlayer?.username ?: "",
                    clickedAdOffer.id
            )

            AlertDialog.Builder(requireContext())
                    .setTitle(R.string.ad_question)
                    .setMessage(R.string.ad_question_description)
                    .setPositiveButton(R.string.button_yes) { _: DialogInterface, _: Int ->
                        AnalyticsUtils().logAdOfferProceed(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedAdOffer.id
                        )

                        loadRewardedVideoAd(clickedAdOffer.adId)
                    }
                    .setNegativeButton(R.string.button_no) { _: DialogInterface, _: Int ->
                        AnalyticsUtils().logAdOfferCancel(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedAdOffer.id
                        )
                    }
                    .setOnCancelListener {
                        AnalyticsUtils().logAdOfferCancel(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedAdOffer.id
                        )
                    }
                    .show()
        }
    }

    override fun onOfferClick(view: View, clickedOffer: Offer) {
        if (context != null && !mSwipeRefreshLayout.isRefreshing && clickedOffer.timeleft == 0 && clickedOffer.price <= mOffersInfo.points) {
            AnalyticsUtils().logOfferDialogOpen(
                    mPlayer?.uuid ?: "",
                    mPlayer?.username ?: "",
                    clickedOffer.id
            )

            AlertDialog.Builder(requireContext())
                    .setTitle(R.string.offer_question)
                    .setMessage(resources.getQuantityString(R.plurals.offer_question_description, clickedOffer.price, clickedOffer.title, clickedOffer.price))
                    .setPositiveButton(R.string.button_yes) { _: DialogInterface, _: Int ->
                        AnalyticsUtils().logOfferProceed(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedOffer.id
                        )

                        redeemOffer(clickedOffer)
                    }
                    .setNegativeButton(R.string.button_no) { _: DialogInterface, _: Int ->
                        AnalyticsUtils().logOfferCancel(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedOffer.id
                        )
                    }
                    .setOnCancelListener {
                        AnalyticsUtils().logOfferCancel(
                                mPlayer?.uuid ?: "",
                                mPlayer?.username ?: "",
                                clickedOffer.id
                        )
                    }
                    .show()
        }
    }

    private fun seekForPlayerUpdates() {
        mViewModel.getPlayer().observe(viewLifecycleOwner, { player: DetailedPlayer? ->
            mPlayer = player
        })
    }

    private fun seekForOffersInfo() {
        showLoadingIndicator()
        mViewModel.setNetworkErrorListener(this)
        mViewModel.getOffersInfo().observe(viewLifecycleOwner, { offersInfo: OffersInfo? ->
            if (offersInfo != null) {
                mOffersInfo = offersInfo
                populateUi()
            }
        })
    }

    fun refreshData() {
        if (isAdded) {
            mSwipeRefreshLayout.isRefreshing = true
            mViewModel.refreshOffersInfo()
        }
    }

    private fun populateUi() {
        showDefaultLayout()
        mCoinsTextView.text = mOffersInfo.points.toString()
        mAdOffersAdapter.setAdOfferList(mOffersInfo.adOffers)
        mOffersAdapter.setOfferList(mOffersInfo.offers, mOffersInfo.points)

        var limitReached = true
        for (adOffer in mOffersInfo.adOffers) {
            if (adOffer.timeleft != -1) {
                limitReached = false
                break
            }
        }
        if (limitReached) {
            mOfferLimitReachedTextView.visibility = View.VISIBLE
        }
        else {
            mOfferLimitReachedTextView.visibility = View.GONE
        }
    }

    private fun loadRewardedVideoAd(adId: String) {
        showLoadingIndicator()
        mRewardedVideoAd.loadAd(adId, AdRequest.Builder().build())
    }

    private fun redeemOffer(offer: Offer) {
        showLoadingIndicator()
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val accessToken = sharedPreferences.getString(getString(R.string.pref_key_access_token), null)
        val context = context

        val serverAPIInterface = retrofitInstance.create(ServerAPIInterface::class.java)
        serverAPIInterface.redeemOffer(ServerAPIClient.API_KEY, offer.id, accessToken!!).enqueue(object : Callback<Void?> {
            override fun onResponse(call: Call<Void?>, response: Response<Void?>) {
                if (isAdded) {
                    AlertDialog.Builder(context!!)
                            .setTitle(R.string.offer_redeemed)
                            .setMessage(R.string.offer_redeemed_description)
                            .setPositiveButton(R.string.button_dismiss, null)
                            .show()
                    refreshData()
                }
            }

            override fun onFailure(call: Call<Void?>, t: Throwable) {
                t.printStackTrace()
            }
        })
    }

    private fun showDefaultLayout() {
        hideLoadingIndicator()
        mNoRanksAvailableTextView.visibility = View.GONE
        mDefaultLayout.visibility = View.VISIBLE
        mNoInternetLayout.visibility = View.GONE
        mServerErrorLayout.visibility = View.GONE
        mAccountBannedLayout.visibility = View.GONE
    }

    private fun showNoRewardsLayout() {
        hideLoadingIndicator()
        mNoRanksAvailableTextView.visibility = View.VISIBLE
        mDefaultLayout.visibility = View.GONE
        mNoInternetLayout.visibility = View.GONE
        mServerErrorLayout.visibility = View.GONE
        mAccountBannedLayout.visibility = View.GONE
    }

    private fun showNoInternetLayout() {
        hideLoadingIndicator()
        mNoRanksAvailableTextView.visibility = View.GONE
        mDefaultLayout.visibility = View.GONE
        mNoInternetLayout.visibility = View.VISIBLE
        mServerErrorLayout.visibility = View.GONE
        mAccountBannedLayout.visibility = View.GONE
    }

    private fun showServerErrorLayout() {
        hideLoadingIndicator()
        mNoRanksAvailableTextView.visibility = View.GONE
        mDefaultLayout.visibility = View.GONE
        mNoInternetLayout.visibility = View.GONE
        mServerErrorLayout.visibility = View.VISIBLE
        mAccountBannedLayout.visibility = View.GONE
    }

    private fun showAccountBannedLayout() {
        hideLoadingIndicator()
        mNoRanksAvailableTextView.visibility = View.GONE
        mDefaultLayout.visibility = View.GONE
        mNoInternetLayout.visibility = View.GONE
        mServerErrorLayout.visibility = View.GONE
        mAccountBannedLayout.visibility = View.VISIBLE
    }

    private fun showLoadingIndicator() {
        if (this::mSwipeRefreshLayout.isInitialized) {
            mSwipeRefreshLayout.isRefreshing = true
        }
    }

    fun hideLoadingIndicator() {
        if (this::mSwipeRefreshLayout.isInitialized) {
            mSwipeRefreshLayout.isRefreshing = false
        }
    }

    fun setRewardedVideoAd(ad: RewardedVideoAd) {
        mRewardedVideoAd = ad
    }

    override fun onNoInternet() {
        showNoInternetLayout()
    }

    override fun onServerError() {
        showServerErrorLayout()
    }
}